docs/site/tutorials/authentication/Authentication-Tutorial.md

```diff
@@ -0,0 +1,974 @@
+
+export class ShoppingApplication extends BootMixin(
+  ServiceMixin(RepositoryMixin(RestApplication)),
+) {
+  constructor(options?: ApplicationConfig) {
+    super(options);
+    // ...
+    registerAuthenticationStrategy(this, JWTAuthenticationStrategy);
+    // ...
+  }
+}
+```
+
+### Creating a Token Service
+
+The token service `JWTService` in
+`loopback4-example-shopping/packages/shopping/src/services/jwt-service.ts`
+implements an **optional** helper
+[TokenService](https://github.com/strongloop/loopback-next/blob/master/packages/authentication/src/services/token.service.ts)
+interface.
+
+```ts
+import {inject} from '@loopback/context';
+import {HttpErrors} from '@loopback/rest';
+import {promisify} from 'util';
+import {TokenService, UserProfile} from '@loopback/authentication';
+import {TokenServiceBindings} from '../keys';
+
+const jwt = require('jsonwebtoken');
+const signAsync = promisify(jwt.sign);
+const verifyAsync = promisify(jwt.verify);
+
+export class JWTService implements TokenService {
+  constructor(
+    @inject(TokenServiceBindings.TOKEN_SECRET)
+    private jwtSecret: string,
+    @inject(TokenServiceBindings.TOKEN_EXPIRES_IN)
+    private jwtExpiresIn: string,
+  ) {}
+
+  async verifyToken(token: string): Promise<UserProfile> {
+    if (!token) {
+      throw new HttpErrors.Unauthorized(
+        `Error verifying token : 'token' is null`,
+      );
+    }
+
+    let userProfile: UserProfile;
+
+    try {
+      // decode user profile from token
+      const decryptedToken = await verifyAsync(token, this.jwtSecret);
+      // don't copy over  token field 'iat' and 'exp', nor 'email' to user profile
+      userProfile = Object.assign(
+        {id: '', name: ''},
+        {id: decryptedToken.id, name: decryptedToken.name},
+      );
+    } catch (error) {
+      throw new HttpErrors.Unauthorized(
+        `Error verifying token : ${error.message}`,
+      );
+    }
+
+    return userProfile;
+  }
+
+  async generateToken(userProfile: UserProfile): Promise<string> {
+    if (!userProfile) {
+      throw new HttpErrors.Unauthorized(
+        'Error generating token : userProfile is null',
+      );
+    }
+
+    // Generate a JSON Web Token
+    let token: string;
+    try {
+      token = await signAsync(userProfile, this.jwtSecret, {
+        expiresIn: Number(this.jwtExpiresIn),
+      });
+    } catch (error) {
+      throw new HttpErrors.Unauthorized(`Error encoding token : ${error}`);
+    }
+
+    return token;
+  }
+}
+```
+
+`JWTService` generates or verifies JWT tokens using the `sign` and `verify`
+functions of [jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken).
+
+It makes use of `jwtSecret` and `jwtExpiresIn` **string** values that are
+injected via the `TokenServiceBindings.TOKEN_SECRET` and the
+`TokenServiceBindings.TOKEN_EXPIRES_IN` binding keys respectively.
+
+The `async generateToken(userProfile: UserProfile): Promise<string>` function
```

Comment:
question: is there a reason the full method definition is used instead of just 'The `generateToken` function'...'?

Reply:
@nabdelgadir 

> question: is there a reason the full method definition is used instead of just 'The generateToken function...'? 

I show the full signature of many methods when I explain them...not just that specify one.

Reply:
yeah, I was wondering why

Reply:
Just a personal preference.